File: 3225-maximum-score-from-grid-operations/solution.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    long long maximumScore(vector<vector<int>>& grid) {
        int n = grid.size();
        // P[j][h] stores the sum of grid[0...h-1][j]
        vector<vector<long long>> P(n, vector<long long>(n + 1, 0));
        for (int j = 0; j < n; ++j) {
            for (int i = 0; i < n; ++i) {
                P[j][i + 1] = P[j][i] + grid[i][j];
            }
        }

        // dp[h_curr][h_prev] = max score for columns 0 to j-1
        // where H_{j} = h_curr and H_{j-1} = h_prev
        vector<vector<long long>> dp(n + 1, vector<long long>(n + 1, 0));

        // Base case: H_{-1} = 0. All H_0 are initialized with 0 score.
        // After j=0 loop, dp will store max score for column 0 given H_1 and H_0.
        
        for (int j = 0; j < n; ++j) {
            vector<vector<long long>> next_dp(n + 1, vector<long long>(n + 1, 0));
            for (int h_j = 0; h_j <= n; ++h_j) {
                // Precompute prefix and suffix maximums to transition H_{j-1} in O(1)
                // prefix_max_dp[k] = max_{h_prev <= k} dp[h_j][h_prev]
                vector<long long> prefix_max_dp(n + 1, 0);
                prefix_max_dp[0] = dp[h_j][0];
                for (int k = 1; k <= n; ++k) prefix_max_dp[k] = max(prefix_max_dp[k - 1], dp[h_j][k]);

                // suffix_max_dp_plus_P[k] = max_{h_prev >= k} (dp[h_j][h_prev] + P[j][h_prev])
                vector<long long> suffix_max_dp_plus_P(n + 2, -1);
                for (int k = n; k >= 0; --k) {
                    suffix_max_dp_plus_P[k] = max(suffix_max_dp_plus_P[k + 1], dp[h_j][k] + P[j][k]);
                }

                for (int h_next = 0; h_next <= n; ++h_next) {
                    long long current_score = 0;
                    if (h_j >= h_next) {
                        // Gain from H_{j-1} only if H_{j-1} > h_j
                        long long val1 = prefix_max_dp[h_j]; 
                        long long val2 = suffix_max_dp_plus_P[h_j + 1] - P[j][h_j];
                        current_score = max(val1, val2);
                    } else {
                        // Gain from H_{j+1} (which is h_next) OR H_{j-1}
                        long long val1 = prefix_max_dp[h_next] + P[j][h_next] - P[j][h_j];
                        long long val2 = suffix_max_dp_plus_P[h_next + 1] - P[j][h_j];
                        current_score = max(val1, val2);
                    }
                    next_dp[h_next][h_j] = current_score;
                }
            }
            dp = move(next_dp);
        }

        // Final answer: H_n is effectively 0
        long long ans = 0;
        for (int h = 0; h <= n; ++h) {
            ans = max(ans, dp[0][h]);
        }
        return ans;
    }
};

File: 3225-maximum-score-from-grid-operations/solution.py
class Solution:
    def maximumScore(self, grid: List[List[int]]) -> int:
        n = len(grid)
        # P[j][h] stores the sum of grid[0...h-1][j]
        P = [[0] * (n + 1) for _ in range(n)]
        for j in range(n):
            for i in range(n):
                P[j][i + 1] = P[j][i] + grid[i][j]

        # dp[h_curr][h_prev] stores max score up to col j-1
        # H_j = h_curr, H_{j-1} = h_prev
        dp = [[0] * (n + 1) for _ in range(n + 1)]

        for j in range(n):
            next_dp = [[0] * (n + 1) for _ in range(n + 1)]
            for h_j in range(n + 1):
                # Precompute prefix max of dp
                prefix_max_dp = [0] * (n + 1)
                curr_max = -1
                for k in range(n + 1):
                    if dp[h_j][k] > curr_max:
                        curr_max = dp[h_j][k]
                    prefix_max_dp[k] = curr_max
                
                # Precompute suffix max of (dp + P[j])
                suffix_max_dp_plus_P = [-1] * (n + 2)
                curr_max_p = -1
                for k in range(n, -1, -1):
                    val = dp[h_j][k] + P[j][k]
                    if val > curr_max_p:
                        curr_max_p = val
                    suffix_max_dp_plus_P[k] = curr_max_p
                
                pj_hj = P[j][h_j]
                for h_next in range(n + 1):
                    if h_j >= h_next:
                        # Case: Previous height or no gain
                        val1 = prefix_max_dp[h_j]
                        val2 = suffix_max_dp_plus_P[h_j + 1] - pj_hj
                        next_dp[h_next][h_j] = val1 if val1 > val2 else val2
                    else:
                        # Case: Next height or previous height
                        pj_hnext = P[j][h_next]
                        val1 = prefix_max_dp[h_next] + pj_hnext - pj_hj
                        val2 = suffix_max_dp_plus_P[h_next + 1] - pj_hj
                        next_dp[h_next][h_j] = val1 if val1 > val2 else val2
            dp = next_dp

        # Result assumes a virtual column n with height 0
        return max(dp[0])